File: DAA/Lab/TW_4_HeapSort/TW4.py
import random
import time


max = 10000

#for generating random numbers
def get_data(array):
    for i in range(max):
        ele = random.randint(0, 80000)
        array.append(ele)


def heapify(array, array_length ,i):
    #initializing the indexes
    parent_index = i
    left_child_index = 2*i +1 
    right_child_index = 2*i +2 

    #checking if left child of root exists and if it is greater than root 
    if left_child_index < array_length and array[parent_index] < array[left_child_index]:
        parent_index = left_child_index
        
    #checking if right child of root exists and if it is greater than root 
    if right_child_index < array_length and array[parent_index] < array[right_child_index]:
        parent_index = right_child_index
    
    #checking if any chnages has been made to parent_index 
    if parent_index != i:
        array[i], array[parent_index] = array[parent_index], array[i] 
        
        heapify(array, array_length, parent_index)


def heapSort(array):
        array_length = len(array)   
        
        #Building the heap for given array
        for i in range((array_length//2 )-1, -1, -1):
            heapify(array, array_length, i)

        #extracting one by one elements
        for i in range(array_length-1, 0, -1):
            array[i], array[0] = array[0], array[i]
            heapify(array, i, 0)



array = []
get_data(array)
start_time = time.time()
heapSort(array)
end_time = time.time()
print(f"time taken is :: {end_time - start_time} seconds")

File: DAA/Lab/TW_4_HeapSort/TW4.c
#include <stdio.h>
#include<stdlib.h>
#include<time.h>
#define MAX 10000
void getdata(int arr[])
{
 int i;
  //generating random numbers
  for(i=0;i<MAX;i++)
  {
     arr[i]=rand();
    }
} 

void heap_bottom_up(int n,int a[])
{
	int i,j,p,k,v,heap;
	p = n/2;
	
	for(i=p;i>0;i--)
	{
		heap = 0;
		k = i;
		v = a[k];
		
		while(!heap&&((2*k)<=n))
		{
			j = 2*k;
			if(j<n)
			{
				if(a[j]<a[j+1])
					j = j+1;
			}
			if(v>=a[j])
				heap = 1;
			else
			{
				a[k] = a[j];
				k = j;
			}
		}
		a[k] = v;
	}
}

void heapsort(int n,int a[])
{
	int i,temp;
	heap_bottom_up(n,a);
	for(i=n;i>1;i--)
	{
		temp = a[1];
		a[1] = a[i];
		a[i] = temp;
		heap_bottom_up(i-1,a);
	}
}

int main()
{
	int i,a[MAX],j,n;
	clock_t start,end;
	double duration;
	//printf("ENTER THE SIZE OF ARRAY : ");
	//scanf("%d",&n);
	
	//printf("ENTER %d ELEMENTS - \n",n);
	//for(i=1;i<=n;i++)
	//	scanf("%d",&a[i]);
	getdata(a);	
	clock_t t;
    t = clock();
    heapsort(MAX,a);
    t = clock() - t;
    double time_taken = ((double)t) / CLOCKS_PER_SEC; // in seconds
    printf("fun() took %f seconds to execute \n", time_taken);
	return 0;
}